### PR TITLE
Library updates to better support removing kubectl deps on kubernetes/kubernetes

### PR DESCRIPTION
**Pull request by pwittrock**
**What this PR does / why we need it**:

**Which issue this PR fixes**: kubernetes/kubectl#50

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by pwittrock** [comment]
cc @kubernetes/sig-cli-pr-reviews

File: pkg/kubectl/cmd/util/factory.go [hunk 1]
```diff
@@ -91,6 +92,10 @@ type ClientAccessFactory interface {
 
 	// ClientSet gives you back an internal, generated clientset
 	ClientSet() (internalclientset.Interface, error)
+
+	// KubernetesClientSetForVersion gives you back an external clientset
+	KubernetesClientSet() (*kubernetes.Clientset, error)
```
Comment (shiywang) [inline pkg/kubectl/cmd/util/factory.go]: KubernetesClientSetForVersion ?

File: pkg/kubectl/cmd/util/factory.go [hunk 2]
```diff
@@ -91,6 +92,10 @@ type ClientAccessFactory interface {
 
 	// ClientSet gives you back an internal, generated clientset
 	ClientSet() (internalclientset.Interface, error)
+
+	// KubernetesClientSetForVersion gives you back an external clientset
```
Comment (shiywang) [inline pkg/kubectl/cmd/util/factory.go]: KubernetesClientSet ?

**Comment by pwittrock** [comment]
please review #50578 instead, this was meant to build on that PR.

**Comment by pwittrock** [comment]
@monopole @apelisse PTAL

**Comment by alexandercampbell** [comment]
@pwittrock please review which PR instead? Your link goes to this PR.

File: pkg/kubectl/cmd/util/clientcache.go [hunk 3]
```diff
@@ -58,6 +59,35 @@ type ClientCache struct {
 	// argument evaluation
 	discoveryClientFactory DiscoveryClientFactory
 	discoveryClient        discovery.DiscoveryInterface
+
+	kubernetesClientCache KubernetesClientCache
+}
+
+// KubernetesClientCache creates a new kubernetes.Clientset one time
+// and then returns the result for all future requests
+type KubernetesClientCache struct {
```
Comment (fabianofranz) [inline pkg/kubectl/cmd/util/clientcache.go]: This type doesn't need to be exposed, just `kubernetesClientCache`.

**Comment by pwittrock** [comment]
PTAL

File: pkg/kubectl/apps/kind_visitor.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+package apps
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// KindVisitor is used with GroupKindElement to call a particular function depending on the
```
Comment (apelisse) [inline pkg/kubectl/apps/kind_visitor.go]: I'm curious how all of this is going to be used :-)

**Pull request merged**